src/redis_db.cc

```diff
@@ -150,31 +159,43 @@ void Database::Keys(std::string prefix, std::vector<std::string> *keys, KeyNumSt
   read_options.snapshot = ss.GetSnapShot();
   read_options.fill_cache = false;
   auto iter = db_->NewIterator(read_options, metadata_cf_handle_);
-  prefix.empty() ? iter->SeekToFirst() : iter->Seek(prefix);
-  for (; iter->Valid(); iter->Next()) {
-    if (!prefix.empty() && !iter->key().starts_with(prefix)) {
-      break;
-    }
-    Metadata metadata(kRedisNone, false);
-    value = iter->value().ToString();
-    metadata.Decode(value);
-    if (metadata.Expired()) {
-      if (stats) stats->n_expired++;
-      continue;
-    }
-    if (stats) {
-      int32_t ttl = metadata.TTL();
-      stats->n_key++;
-      if (ttl != -1) {
-        stats->n_expires++;
-        if (ttl > 0) ttl_sum += ttl;
+
+  while (true) {
+    ns_prefix.empty() ? iter->SeekToFirst() : iter->Seek(ns_prefix);
+    for (; iter->Valid(); iter->Next()) {
+      if (!prefix.empty() && !iter->key().starts_with(ns_prefix)) {
+        break;
+      }
+      Metadata metadata(kRedisNone, false);
+      value = iter->value().ToString();
+      metadata.Decode(value);
+      if (metadata.Expired()) {
+        if (stats) stats->n_expired++;
+        continue;
+      }
+      if (stats) {
+        int32_t ttl = metadata.TTL();
+        stats->n_key++;
+        if (ttl != -1) {
+          stats->n_expires++;
+          if (ttl > 0) ttl_sum += ttl;
+        }
+      }
+      if (keys) {
+        ExtractNamespaceKey(iter->key(), &ns, &user_key, storage_->IsClusterEnabled());
+        keys->emplace_back(user_key);
       }
     }
-    if (keys) {
-      ExtractNamespaceKey(iter->key(), &ns, &user_key);
-      keys->emplace_back(user_key);
-    }
+
+    if (!storage_->IsClusterEnabled()) break;
+    if (prefix.empty()) break;
+    if (++slot_id >= HASH_SLOTS_SIZE) break;
+
+    ComposeNamespaceKey(namespace_, "", &ns_prefix);
+    PutFixed16(&ns_prefix, slot_id);
+    ns_prefix.append(prefix);
   }
+
```

Comment:
It Seems uses do...while sentence would be more clear, like below:
```
do {
} while(storage_->IsClusterEnabled() && ++slot_id < HASH_SLOTS_SIZE)
```

Reply:
it is not easy to use this way, because we may need to init some variables.